kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -194,6 +196,24 @@ public Transaction build(Engine engine) {
       TableFeatures.validateKernelCanWriteToTable(protocol, metadata, table.getPath(engine));
     }
 
+    /* Column metadata must not be updated in case of schema evolution since field IDs must be
+    preserved */
+    if (schema.isPresent() && !isNewTable) {
+      SchemaUtils.validateUpdatedSchema(
+          snapshot.getSchema(),
+          schema.get(),
+          snapshot.getPartitionColumnNames(),
+          snapshot.getMetadata());
+      shouldUpdateMetadata = true;
+      metadata = metadata.withNewSchema(schema.get());
+    } else if (isNewTable || shouldUpdateMetadata) {
+      metadata =
```

Comment:
This essentially lifts the change out of TransactionImpl, similar to what was done in https://github.com/delta-io/delta/pull/4199/files#diff-e18199e992a0835a88ce135292bb9d3b6b7cc004406108d6eefd5981bdd06354L319

If @vkorukanti gets his changes in first, I'll need to rebase this so that we update the column mapping metadata under the same conditions as his PR but I think as of the current behavior this condition is right.

This gets rid of the boolean I originally had in `TransactionImpl`